### PR TITLE
Modify runlength usage - options for using 1D and Nd versions of code

### LINKED ISSUE #217: run length calculation optimization
Indices using n-Dim runlength #139 functions on small spatial  subsets is inefficient.
see #213 for details

We can potentially attempt to determine when to use 1d/ufunc method vs n-Dim method based on grid size

ideally this would be dealt with directly in run_length.py... 
edit - more detailed discussion here : https://github.com/Ouranosinc/xclim/issues/213#issuecomment-497706760

### PR DESCRIPTION
**Pull request by tlogan2000**
* **Please check if the PR fulfills these requirements**
- [x] This PR addresses an already opened issue (for bug fixes / features)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] HISTORY.rst has been updated
- [ ] bumpversion (minor / major) has been called
- [ ] Tags have been pushed

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, etc.)
Modify runlength calculation usage to optimize speed #217

* **Does this PR introduce a breaking change?** (Has there been an API change?)
no

* **Other information**:

first attempt to fix #217

### CODE REVIEW

**Comment by tlogan2000** [comment]
@huard this is a first attempt based on test results in #217.  Opinions on whether this is a reasonable approach?

**Comment by tlogan2000** [comment]
test are failing ... hold off for a bit

File: xclim/run_length.py [hunk 1]
```diff
@@ -7,6 +7,16 @@
 from warnings import warn
 
 logging.captureWarnings(True)
+npts_opt = 9000
+
+
+def get_npts(da):
```
Comment (huard) [inline xclim/run_length.py]: Docstring

File: xclim/run_length.py [hunk 2]
```diff
@@ -41,7 +51,7 @@ def rle(da, dim='time', max_chunk=1000000):
     return d
 
 
-def longest_run(da, dim='time'):
+def longest_run(da, dim='time', ufunc_1Dim='auto'):
```
Comment (huard) [inline xclim/run_length.py]: I think pep suggest lowercase variable names.

File: xclim/run_length.py [hunk 3]
```diff
@@ -120,22 +159,34 @@ def first_run(da, window, dim='time'):
...
-        da.swap_dims({dim: 'time'})
-    da = da.astype('int')
-    i = xr.DataArray(np.arange(da[dim].size), dims=dim).chunk({'time': 1})
-    ind = xr.broadcast(i, da)[0].chunk(da.chunks)
-    wind_sum = da.rolling(time=window).sum()
-    out = ind.where(wind_sum >= window).min(dim=dim) - (
-        window - 1)  # remove window -1 as rolling result index is last element of the moving window
+    if ufunc_1Dim == 'auto':
+        npts = get_npts(da)
+        ufunc_1Dim = bool(npts <= npts_opt)
```
Comment (huard) [inline xclim/run_length.py]: isn't this a bool anyway ?

File: xclim/run_length.py (hunk 1 above)
Comment (tlogan2000) [reply xclim/run_length.py]: I agree,  I do not completely understand why this doesn't scale efficiently.  I'm not sure we have any other explicit uses of ufuncs(?) ...

**Comment by tlogan2000** [comment]
> I think it's weird that we need to do this. Do other ufuncs suffer from the same issues ?
ufunc scaling is actually close to perfectly linear (blue line) with the number of points.  I think the extra steps in the n-dim version (index variables, bfills etc etc) are maybe simply a bit too much overhead to be worth it when grid dimensions are small.

image

**Comment by huard** [comment]
rle_1d computes three different things: 
- the values taken by each run
- each run's length
- the starting position

I think we could probably speed that function by assuming we pass a boolean and look only for runs of True.

**Pull request merged**